h2o-persist-s3/src/main/java/water/persist/PersistS3.java

```diff
@@ -25,35 +28,74 @@
+  private static final Pattern URL_CREDENTIALS_PATTERN = Pattern.compile("s3://(.[^:]*):{0,1}(.[^@]*)@{1}(.*)");
 
+  /**
+   * Creates an Amazon S3 client using default chain of credential providers, without incorporating access key ID and
+   * secret access key. This kind of {@link AmazonS3} client is constructed for any purpose but 'GET' operations,
+   * where in-url credentials are not supported.
+   *
+   * @return An instance of {@link AmazonS3Client}
+   */
   public static AmazonS3 getClient() {
-    if( _s3 == null ) {
-      synchronized( _lock ) {
-        if( _s3 == null ) {
-          try {
-            H2OAWSCredentialsProviderChain c = new H2OAWSCredentialsProviderChain();
-            ClientConfiguration cc = s3ClientCfg();
-            _s3 = configureClient(new AmazonS3Client(c, cc));
-          } catch( Throwable e ) {
-            e.printStackTrace();
-            StringBuilder msg = new StringBuilder();
-            msg.append(e.getMessage() + "\n");
-            msg.append("Unable to load S3 credentials.");
-            throw new RuntimeException(msg.toString());
-          }
+    return getClient(null, null);
+  }
+
+  /**
+   * Creates an Amazon S3 client using default chain of credential providers, enhanced with static credentials providers
+   * based on accessKeyId and accessSecretKey pair - if provided.
+   *
+   * Any time accessKeyId and accessSecretKey arguments are provided, their value is hashed and compared to those used to
+   * create existing client. If the credentials are not equal (comparison includes null values), new client is constructed.
+   *
+   * @param accessKeyId Secret access Key ID. If null, treated as not provided and not used in the credential providers chain.
+   * @param accessSecretKey Secret access key. If null, treated as not provided and not used in the credential providers chain.
+   * @return An instance of {@link AmazonS3}, either newly constructed or existing instance
+   */
+  public static AmazonS3 getClient(final String accessKeyId, final String accessSecretKey) {
+
+    final byte[] digest;
+    // When users specify credentials in URL, these credentials might be different than those provided in configuration
+    // or provided in previous URL used. If the credentials are different, we must build a new client with new credentials.
+    // Not providing credentials also correctly results in new client construction (users may want to continue with values from config).
+    // Storing these values in memory for comparison is potentially dangerous. Instead, a hash of those two values is stored.
+    if (accessKeyId != null && accessSecretKey != null) {
+      StringBuilder keyDigestBuilder = new StringBuilder(accessKeyId);
+      keyDigestBuilder.append(accessSecretKey);
+      digest = _messageDigest.digest(keyDigestBuilder.toString().getBytes());
+    } else {
+      digest = null;
+    }
+
+    synchronized (_lock) {
+      if (_s3 == null || digest != _credentialsDigest) {
```

Comment:
`_credentialsDigest ` is never updated and stays `null` in this PR.

Reply:
I know. Latest patch not applied. WIP.